sklearn/cluster/_birch.py

```diff
@@ -654,11 +653,10 @@ def predict(self, X):
         """
         check_is_fitted(self)
         X = self._validate_data(X, accept_sparse="csr", reset=False)
-        kwargs = {"Y_norm_squared": self._subcluster_norms}
```

Comment:
> Precomputing those norms and reusing them for several consecutive predict calls might still be beneficial from a perf point of view.

Yes, definitely.

> Could you please run a quick benchmarks where Birch is fit once on a dataset and then predict is called several times and compare the prediction speed in main vs this PR on a regular laptop or a server with 4 cores or so?

Will do.

> Being able to pass precomputed Y_norm_squared to the new implementation of fast_sqeuclidean would probably still be a good idea (for instance to also speed-up a KNN predict method where the model is fit once on a large training set and then makes predictions on many small prediction sets) but I am not sure how complex it would be to implement as part of this PR.

I suggest implementing it in another PR because this will necessitates some boilerplate code we probably do not know to introduce it this PR large diff. WDYT?